src/world/Spell/Spell.cpp

```diff
@@ -1376,6 +1809,106 @@ void Spell::sendCastResult(Player* caster, uint8_t castCount, SpellCastResult re
     caster->sendCastFailedPacket(getSpellInfo()->getId(), result, castCount, parameter1, parameter2);
 }
 
+int32_t Spell::getFullCastTime() const
+{
+    return m_castTime;
+}
+
+int32_t Spell::getCastTimeLeft() const
+{
+    return m_timer;
+}
+
+uint32_t Spell::calculatePowerCost() const
+{
+    // Zero for non-unit casters
+    if (u_caster == nullptr)
+        return 0;
+
+    int32_t powerCost = 0;
+    if (getSpellInfo()->getAttributesEx() & ATTRIBUTESEX_DRAIN_WHOLE_POWER)
+    {
+        if (getSpellInfo()->getPowerType() == POWER_TYPE_HEALTH)
+            powerCost = static_cast<int32_t>(u_caster->getHealth());
+#if VERSION_STRING < WotLK
```

Comment:
move multiversion specific definitions to its own function e.g. bool isValidPowerType(PowerType type), to simplyfie code.